### PR TITLE
add cut-release github workflow

### PR DESCRIPTION
**Pull request by lrfurtado**
this pull request attempts to address issue #140

### CODE REVIEW

**Comment by andikleen** [comment]
Thanks.

Hmm, I must say I'm a little uncomfortable with all the dependencies on other people's packages in that action. What happens if they change something? Normally numactl avoids dependencies.

Can't we just write the actions directly? Nothing should be very complicated.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:50]: Should we pin this to [marvinpinto/actions@v1.2.1](https://github.com/marvinpinto/actions/releases/tag/v1.2.1) perhaps?

That's currently the [latest versioned release of that action](https://github.com/marvinpinto/action-automatic-releases/releases).

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:17]: Create a separate step at the beginning to install packages with apt-get and then use it to install everything that you need. It's more efficient that way.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:20]: Can you also install the `fakeroot` package with apt-get and then use `fakeroot make distcheck`? This will produce tarballs with ownership `root/root` (0/0) instead of the current `runner/docker` (1001/122) which might produce a cleaner tree when unpacked as root at a target machine.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:40]: Let's use a commit message of `Release numactl {{ steps.bump-version-file.outputs.NEW_VERSION }}` here.

See dd6de072c92c892a86e18c0fd0dfa1ba57a9a05d (commit for bumping up version to 2.0.14) for an example.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:47]: Let's use a tag description of `Release numactl {{ steps.bump-version-file.outputs.NEW_VERSION }}` here.

See https://github.com/numactl/numactl/releases/tag/v2.0.14 for an example.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:7]: How about "Version (leave blank for auto-increment)" as the label for the field?

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:36]: I think you're missing a step to push the commit (and tag?) to the master/main branch of the repository.

At least I looked at [your test releases](https://github.com/lrfurtado/numactl/releases) and they don't seem to be created on a commit that's bumping the version up. Can you please confirm that this works as expected?

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:53]: Hmmm, isn't the "Create Tag" above redundant if we're going to have the release auto-create the tag?

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:55]: Perhaps we *should* mark this as a draft (or perhaps pre-release, though I think draft sounds more appropriate here) so that a maintainer has a chance to do some final checks before making it fully available?

You can consider making that configurable, perhaps through an input checkbox, checked by default, for "Make this a draft release", up above.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:58]: I believe this should always be `numactl-$(cat VERSION).tar.gz`, so maybe let's be more specific here and simply use `numactl-*.tar.gz` as a wildcard to match these? You can skip the `*tgz` one.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:31]: It would be good to compare that new version is larger than older version here. You can use `pysemver compare` for that purpose, check the output, if it's -1 or 0 it means the first argument is smaller or they're equal, if it's 1 it means the second argument is the newer semver (which is what we want).

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:27]: How about `NEW_VERSION=$(pysemver bump patch "$VERSION")` and then you can use common logic using `$NEW_VERSION`.

You might want to store `VERSION=$(cat VERSION)` at the start to be able to do comparisons, etc.

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:23]: You can use `set -eu -o pipefail` to ensure failure in any command will make this break.

**Review by filbranden — CHANGES_REQUESTED** [review]
This looks awesome already!

I have some comments and follow ups, but I'm already pretty happy with how this is looking, thank you so much for contributing!

**Comment by filbranden** [comment]
@andikleen 

> Hmm, I must say I'm a little uncomfortable with all the dependencies on other people's packages in that action. What happens if they change something? 

I understand this is a little bit awkward but it's pretty much standard practice with GitHub Actions. Furthermore, it seems most of them are already pinned to specific git tags (I mean, technically you could move those, but one is not supposed to) and the one that was referring to `latest` I left a comment suggesting pinning to a specific tag as well.

> Normally numactl avoids dependencies.

These are not really dependencies of numactl itself, it's only about automating cutting a new release here, no one building or using numactl will depend on any of these, none of the numactl code depends or builds against or using any of this.

> Can't we just write the actions directly? Nothing should be very complicated.

Actually managing the GitHub API and creating/pushing git commits can be quite complicated in terms of managing the tokens and secrets to get permissions to do that. I think reusing code to perform these actions is actually a pretty reasonable and standard thing to do.

I left specific comments and it seems we're missing one part which is pushing the commit with the version to trunk, but overall I'm very happy with this improvement to the release pipeline!

File: .github/workflows/cut-release.yml
Comment (lrfurtado) [reply .github/workflows/cut-release.yml:36]: you are not seing them because I keep force pushing the latest squased version of my code.

File: .github/workflows/cut-release.yml
Comment (lrfurtado) [reply .github/workflows/cut-release.yml:53]: you are right it is not needed I will take it out.

File: .github/workflows/cut-release.yml
Comment (filbranden) [reply .github/workflows/cut-release.yml:36]: It’s a pity you started this PR with your “master” branch, since then your experiments end up showing up at the PR and you need to clean them up every time.

Consider renaming your “master” branch on your GitHub repo to another name. You can do that (while keeping this PR open) by using the following procedure:
https://github.blog/changelog/2021-01-19-support-for-renaming-an-existing-branch/

Then, you can use your repo’s “master” (or “main”) for experiments, and you can link to commits and tags there from this PR so that we can look at how it shows when you test it.

I’ll probably have some time at the computer this afternoon (on the phone right now) so I’ll take your PR for a spin.

Thanks for all your work so far! This is really useful!

Comment (lrfurtado) [reply .github/workflows/cut-release.yml:36]: The problem was if the code was not on main/master the workflow would not get picked up for testing , but I will play with the branch names.

Comment (filbranden) [reply .github/workflows/cut-release.yml:36]: Yes, sure. You can handle that by merging your feature branch into your repo’s master/main to test the feature.

**Comment by andikleen** [comment]
Okay I guess we can use it with the requested changes

File: .github/workflows/cut-release.yml
Comment (filbranden) [reply .github/workflows/cut-release.yml:36]: I checked this PR in [my repo](https://github.com/filbranden/numactl) and I still see that the release/tag is being created not on the new "Release 2.0.16" commit, but on the commit that precedes it.

See for example [my tags](https://github.com/filbranden/numactl/tags):

Screen Shot 2022-09-25 at 12 25 49 AM

Or [my releases](https://github.com/filbranden/numactl/releases) (after changing it from draft to published):

Screen Shot 2022-09-25 at 12 27 12 AM

Looking at [my commits](https://github.com/filbranden/numactl/commits/main), it's clear that the release commit is being created (in my case, it's f63cf6f635eb972ff3ef0830e9fa7079ea880da3 on top of your 7d4f68bbc6e1634e036ef69c45a47902f013545a), but the tag is being created incorrectly.

The tarball looks good, it has the updated VERSION contents and unpacking and building it works as expected, so all is good on that front!

Can you please take one more look into why this is not picking up that last pushed commit when the tag is being created? Thanks!

File: .github/workflows/cut-release.yml
Comment (filbranden) [inline .github/workflows/cut-release.yml:49]: I'm getting `Warning: Unexpected input(s) 'release_name', valid inputs are ['repo_token', 'automatic_release_tag', 'draft', 'prerelease', 'title', 'files']`

File: .github/workflows/cut-release.yml
Comment (filbranden) [reply .github/workflows/cut-release.yml:48]: So I've been looking at this and it seems this is creating an unsigned tag (see `git help tag` for the `-s` option).

The git "signed" tags are usually better, they create an extra tag object and not just a "pointer" to a commit.

See the difference here:

```
$ git show v2.0.13
tag v2.0.13
Tagger: Filipe Brandenburger <filbranden@gmail.com>
Date:   Fri Aug 30 01:16:08 2019 -0700

Release numactl 2.0.13

commit 5d9f16722e3df49dc618a9f361bd482559695db7 (tag: v2.0.13)
Author: Filipe Brandenburger <filbranden@gmail.com>
Date:   Fri Aug 30 01:11:16 2019 -0700

    Release numactl 2.0.13

    Signed-off-by: Filipe Brandenburger <filbranden@gmail.com>

diff --git a/configure.ac b/configure.ac
index 572b53d..b615f1f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1,5 +1,5 @@
 AC_PREREQ([2.64])
-AC_INIT([numactl], [2.0.12])
+AC_INIT([numactl], [2.0.13])
```

And:

```
$ git show v2.0.16
commit 7d4f68bbc6e1634e036ef69c45a47902f013545a (HEAD -> main, tag: v2.0.16, lrfurtado/master, lrfurtado/cutrelease, lrfurtado)
Author: Luciano Furtado <lrfurtado@yahoo.com.br>
Date:   Sat Sep 24 17:06:13 2022 +0000

    add cut-release github workflow

diff --git a/.github/workflows/cut-release.yml b/.github/workflows/cut-release.yml
new file mode 100644
index 0000000..277df58
--- /dev/null
+++ b/.github/workflows/cut-release.yml
@@ -0,0 +1,53 @@
+name: "cut-release"
...
```

Ignoring that the tag is pointing at the wrong commit, you can also see it's not a signed tag. Ideally we should use one. Perhaps the separate "Create Tag" action would be able to provide that correctly?

**Review by filbranden — CHANGES_REQUESTED** [review]

File: .github/workflows/cut-release.yml
Comment (filbranden) [reply .github/workflows/cut-release.yml:48]: Looking at `EndBug/add-and-commit`, it seems it supports [tagging the created commit](https://github.com/EndBug/add-and-commit#tagging), and it seems you can pass it arguments to the `git tag` command that might include the `-a` flag (for an annotated tag) including a `-m` for a message for the annotated tag itself. Can you take a look at that?

File: .github/workflows/cut-release.yml
Comment (filbranden) [reply .github/workflows/cut-release.yml:36]: I've done some experimentation and didn't find a way to solve this by doing both in the same action.

I think you actually need to split this into two separate actions.

One to just update the VERSION file and create a new commit with an annotated tag and push that to the main branch on the repository.

Then another separate one that gets triggered when a tag is pushed, and then it does the build and creates the GitHub release. The second one has both the `make distcheck` part and the marvinpinto/action-automatic-releases action. See the ["when tags are pushed to repository"](https://github.com/marvinpinto/action-automatic-releases#create-a-new-github-release-when-tags-are-pushed-to-the-repository) example, but in short you probably want this as a trigger for the action that builds a tarball and creates a GitHub release:

```
on:
  push:
    tags:
      - "v*"
```

**Comment by andikleen** [comment]
Do you guys think we can resolve this soon? Otherwise I will do a manual release to fix the ABI break issue. Thanks.

**Comment by filbranden** [comment]
> Do you guys think we can resolve this soon? Otherwise I will do a manual release to fix the ABI break issue. Thanks.

Sure! I think @lrfurtado is unavailable but I can pick it up from here.

We have discussed the approach offline, we'll split this into two separate actions, one to push the version file and tag and the second one to build and create the release.

I'll update this PR to do the latter now.

**Comment by filbranden** [comment]
Alright, I just tested the simplified workflow in my local repo and I'm happy with the results.

Here's the action run:
https://github.com/filbranden/numactl/actions/runs/3147975679/jobs/5118020578

And the generated release v2.0.16:
https://github.com/filbranden/numactl/releases/tag/v2.0.16

Tarball `numactl-2.0.16.tar.gz` attached to the release also looks good.

I'll go ahead and squash commits and merge this PR into trunk, so at least this part that is very GitHub-specific gets automated, also it includes the build of the release tarball which is something that was missed on the last release.

**Comment by filbranden** [comment]
Since the first part of the release bump is not ready yet, the procedure to release version v2.0.16 for now will be:

```
$ echo 2.0.16 >VERSION
$ git commit -am 'Release numactl 2.0.16'
$ git tag -a v2.0.16 -m 'Release numactl 2.0.16'
$ git push origin v2.0.16
```

I'm happy to execute those and push a new tag to release v2.0.16. @andikleen ok for me to proceed with that once this PR is merged?

**Pull request merged**
